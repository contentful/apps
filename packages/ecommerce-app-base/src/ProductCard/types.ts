import { Product } from '../types';

export type ExternalResource = Product;

// the variant states of ProductCard, for now, labeled by location
export type ProductCardType = 'field' | 'dialog';

export type ExternalResourceError = {
  error: string;
  errorMessage: string;
  errorStatus: number;
};
