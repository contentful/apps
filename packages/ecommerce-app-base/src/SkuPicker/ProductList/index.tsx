import React from 'react';
import tokens from '@contentful/forma-36-tokens';
import { css } from '@emotion/css';
import { Product } from '../../interfaces';
import { ProductListItem } from './ProductListItem';

export interface Props {
  products: Product[];
  selectProduct: (sku: string) => void;
  selectedSKUs: string[];
  skuLabel?: string;
  readableIdentifierLabel?: string;
}

const styles = {
  productList: css({
    display: 'flex',
    flexWrap: 'wrap',
    marginLeft: `-${tokens.spacingS}`,
    marginRight: `-${tokens.spacingS}`
  })
};

export const ProductList = ({
  selectProduct,
  selectedSKUs,
  products,
  skuLabel,
  readableIdentifierLabel,
}: Props) => (
  <div className={styles.productList}>
    {products.map(product => (
      <ProductListItem
        key={product.id}
        product={product}
        isSelected={selectedSKUs.includes(product.sku)}
        selectProduct={selectProduct}
        skuLabel={skuLabel}
        readableIdentifierLabel={readableIdentifierLabel}
      />
    ))}
  </div>
);
