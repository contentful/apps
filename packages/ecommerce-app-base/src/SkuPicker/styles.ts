import { css } from 'emotion';
import tokens from '@contentful/f36-tokens';

const HEADER_HEIGHT = 114;

const STICKY_HEADER_BREAKPOINT = 900;

function makeBodyStyle() {
  // Modal window height dynamically calculated in pixels.
  // Usage of css relative units was causing a bug with
  // multiple resizings of the modal window.
  const thirtyPerc = window.outerHeight * 0.3;
  const height = window.outerHeight - thirtyPerc - HEADER_HEIGHT;

  return css({
    height,
    padding: `${tokens.spacingL} ${tokens.spacingL} 0 ${tokens.spacingL}`,

    [`@media screen and (min-height: ${STICKY_HEADER_BREAKPOINT}px)`]: {
      padding: `calc(${tokens.spacingL} + ${HEADER_HEIGHT}px)  ${tokens.spacingL} 0 ${tokens.spacingL}`,
    },
  });
}

export const styles = {
  header: css({
    borderBottom: `1px solid ${tokens.gray300}`,
    display: 'flex',
    justifyContent: 'space-between',
    alignItems: 'flex-start',
    padding: tokens.spacingL,

    [`@media screen and (min-height: ${STICKY_HEADER_BREAKPOINT}px)`]: {
      backgroundColor: 'white',
      position: 'fixed',
      top: 0,
      zIndex: 1,
      width: `calc(100% - 2rem)`,
    },
  }),
  body: makeBodyStyle(),
  total: css({
    fontSize: tokens.fontSizeS,
    color: tokens.gray600,
    display: 'block',
    marginTop: tokens.spacingS,
  }),
  saveBtn: css({
    marginRight: tokens.spacingM,
  }),
  paginator: css({
    margin: `${tokens.spacingM} auto ${tokens.spacingL} auto`,
    textAlign: 'center',
  }),
  leftsideControls: css({
    position: 'relative',
    zIndex: 0,
    svg: css({
      zIndex: 1,
      position: 'absolute',
      top: '10px',
      left: '10px',
    }),
    input: css({
      paddingLeft: '35px',
    }),
  }),
  rightsideControls: css({
    justifyContent: 'flex-end',
    flexGrow: 1,
    display: 'flex',
  }),
  loadMoreButton: css({
    marginTop: tokens.spacingXs,
  }),
};
