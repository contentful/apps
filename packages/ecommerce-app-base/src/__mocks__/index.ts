export { products, productsList } from './products';
export { makeSdkMock } from './mockSdk';
export { mockContentTypes } from './mockContentTypes';
