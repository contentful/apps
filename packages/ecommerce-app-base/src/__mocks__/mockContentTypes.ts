export const mockContentTypes = [
  {
    sys: { id: 'ct1' },
    name: 'CT1',
    fields: [
      { id: 'product_x', name: 'Product X', type: 'Symbol' },
      { id: 'y', name: 'Y', type: 'Object' },
    ],
  },
  {
    sys: { id: 'ct2' },
    name: 'CT2',
    fields: [
      { id: 'foo', name: 'FOO', type: 'Text' },
      { id: 'z', name: 'Z', type: 'Array', items: { type: 'Symbol' } },
    ],
  },
  {
    sys: { id: 'ct3' },
    name: 'CT3',
    fields: [
      { id: 'bar', name: 'BAR', type: 'Object' },
      { id: 'baz', name: 'BAZ', type: 'Object' },
      { id: 'product_d', name: 'Product D', type: 'Array', items: { type: 'Symbol' } },
      { id: 'product_a', name: 'Product A', type: 'Symbol' },
    ],
  },
];
