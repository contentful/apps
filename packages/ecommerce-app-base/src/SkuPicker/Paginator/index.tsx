import React from 'react';
import clamp from 'lodash/clamp';
import range from 'lodash/range';
import { doubleChevron } from '../iconsInBase64';
import { styles } from './styles';

import { Button } from '@contentful/f36-components';

import { ChevronLeftIcon, ChevronRightIcon } from '@contentful/f36-icons';

export interface Props {
  activePage: number;
  className?: string;
  pageCount: number;
  setActivePage: (page: number) => void;
}

export function getPagesRange(page: number, total: number, neighboursCount = 2): number[] {
  const PAGINATOR_RANGE = neighboursCount * 2;

  if (total <= PAGINATOR_RANGE) {
    // Total amount of pages are less than the possible paginator range
    return range(0, total);
  }
  if (page <= neighboursCount) {
    // Active page is at the start of the paginator page count
    return range(0, PAGINATOR_RANGE + 1);
  }
  if (page > total - neighboursCount) {
    // Active page is at the end of the paginator page count
    return range(total - PAGINATOR_RANGE - 1, total);
  }
  // Active page is in the middle of the paginator count
  return range(page - neighboursCount - 1, page + neighboursCount);
}

export function Paginator(props: Props) {
  const { className, pageCount, setActivePage } = props;
  const activePage = clamp(props.activePage, 1, pageCount);
  const hasOnlyOnePage = pageCount === 1;
  const activePageIsAtPaginatorStart = activePage === 1;
  const activePageIsAtPaginatorEnd = activePage === pageCount;

  return (
    <div className={className}>
      <Button
        className={styles.button}
        variant="secondary"
        isDisabled={hasOnlyOnePage || activePageIsAtPaginatorStart}
        onClick={() => setActivePage(1)}
      >
        <img className={styles.chevronLeft} src={doubleChevron} alt="right" />
      </Button>
      <Button
        startIcon={<ChevronLeftIcon />}
        className={styles.button}
        variant="secondary"
        isDisabled={hasOnlyOnePage || activePageIsAtPaginatorStart}
        onClick={() => setActivePage(activePage - 1)}
      />
      {getPagesRange(activePage, pageCount).map((pageIndex) => {
        const page = pageIndex + 1;
        return (
          <Button
            onClick={() => setActivePage(page)}
            className={styles.button}
            variant={page === activePage ? 'primary' : 'secondary'}
            testId={page === activePage ? 'active' : `inactive-${page}`}
            key={pageIndex}
          >
            {page}
          </Button>
        );
      })}
      <Button
        startIcon={<ChevronRightIcon />}
        variant="secondary"
        className={styles.button}
        isDisabled={hasOnlyOnePage || activePageIsAtPaginatorEnd}
        onClick={() => setActivePage(activePage + 1)}
      />
      <Button
        variant="secondary"
        className={styles.button}
        isDisabled={hasOnlyOnePage || activePageIsAtPaginatorEnd}
        onClick={() => setActivePage(pageCount)}
      >
        <img className={styles.chevronRight} src={doubleChevron} alt="right" />
      </Button>
    </div>
  );
}
