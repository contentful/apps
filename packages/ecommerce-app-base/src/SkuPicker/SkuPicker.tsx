import React, { Component } from 'react';
import get from 'lodash/get';
import clamp from 'lodash/clamp';
import debounce from 'lodash/debounce';
import { DialogExtensionSDK } from '@contentful/app-sdk';
import { ProductList } from './ProductList';
import { Paginator } from './Paginator';
import {
  MakeSaveBtnTextFn,
  Pagination,
  Product,
  ProductPreviewsFn,
  ProductsFn,
} from '../interfaces';
import { ProductSelectionList } from './ProductSelectionList';
import { styles } from './styles';
import { mapSort } from '../utils';

import { Button, Icon, TextInput } from '@contentful/f36-components';

import { SearchIcon } from '@contentful/f36-icons';

export interface Props {
  sdk: DialogExtensionSDK;
  fetchProductPreviews: ProductPreviewsFn;
  fetchProducts: ProductsFn;
  searchDelay?: number;
  skuType?: string;
  makeSaveBtnText?: MakeSaveBtnTextFn;
  hideSearch?: boolean;
}

interface State {
  activePage: number;
  search: string;
  pagination: Pagination;
  products: Product[];
  selectedProducts: Product[];
  selectedSKUs: string[];
}

const DEFAULT_SEARCH_DELAY = 250;

function defaultGetSaveBtnText(selectedSKUs: string[]): string {
  switch (selectedSKUs.length) {
    case 0:
      return 'Save products';
    case 1:
      return 'Save 1 product';
    default:
      return `Save ${selectedSKUs.length} products`;
  }
}

export class SkuPicker extends Component<Props, State> {
  state: State = {
    activePage: 1,
    search: '',
    pagination: {
      count: 0,
      limit: 0,
      offset: 0,
      total: 0,
    },
    products: [],
    selectedProducts: [],
    selectedSKUs: get(this.props, ['sdk', 'parameters', 'invocation', 'fieldValue'], []),
  };

  setSearchCallback: () => void;

  constructor(props: Props) {
    super(props);
    this.setSearchCallback = debounce(() => {
      this.setActivePage(1);
    }, this.props.searchDelay || DEFAULT_SEARCH_DELAY);
  }

  componentDidMount() {
    this.updateProducts();
    this.updateSelectedProducts();
  }

  setSearch = (search: string) => {
    this.setState({ search }, this.setSearchCallback);
  };

  updateProducts = async () => {
    try {
      const {
        activePage,
        pagination: { limit },
        search,
      } = this.state;
      const offset = (activePage - 1) * limit;
      const fetched = await this.props.fetchProducts(search, { offset });
      // If the request has been cancelled because a new one has been launched
      // then fetchProducts will return null
      if (fetched && fetched.pagination && fetched.products) {
        this.setState({ pagination: fetched.pagination, products: fetched.products });
      }
    } catch (error) {
      this.props.sdk.notifier.error('There was an error fetching the product list.');
    }
  };

  updateSelectedProducts = async () => {
    try {
      const { selectedSKUs } = this.state;
      const { sdk, skuType, fetchProductPreviews } = this.props;
      const config = sdk.parameters.installation;
      const selectedProductsUnsorted = await fetchProductPreviews(selectedSKUs, config, skuType);
      const selectedProducts = mapSort(selectedProductsUnsorted, selectedSKUs, 'sku');
      this.setState({ selectedProducts });
    } catch (error) {
      this.props.sdk.notifier.error(
        'There was an error fetching the data for the selected products.'
      );
    }
  };

  loadMoreProducts = async () => {
    const { pagination, products } = await this.props.fetchProducts(this.state.search);
    this.setState((oldState) => ({ pagination, products: [...oldState.products, ...products] }));
  };

  setActivePage = (activePage: number) => {
    const { pagination } = this.state;
    const pageCount = Math.ceil(pagination.total / pagination.limit);
    this.setState({ activePage: clamp(activePage, 1, pageCount) }, () => {
      this.updateProducts();
    });
  };

  selectProduct = (sku: string) => {
    const { fieldType } = this.props.sdk.parameters.invocation as Record<string, any>;
    const onlyOneProductCanBeSelected = fieldType === 'Symbol';

    if (this.state.selectedSKUs.includes(sku)) {
      this.setState(
        ({ selectedSKUs }) => ({
          selectedSKUs: selectedSKUs.filter((productSku) => productSku !== sku),
        }),
        () => this.updateSelectedProducts()
      );
    } else {
      this.setState(
        ({ selectedSKUs }) => ({
          selectedSKUs: onlyOneProductCanBeSelected ? [sku] : [...selectedSKUs, sku],
        }),
        () => this.updateSelectedProducts()
      );
    }
  };

  render() {
    const { search, pagination, products, selectedProducts, selectedSKUs } = this.state;
    const { makeSaveBtnText = defaultGetSaveBtnText, skuType, hideSearch = false } = this.props;
    const infiniteScrollingPaginationMode = 'hasNextPage' in pagination;
    const pageCount = Math.ceil(pagination.total / pagination.limit);

    return (
      <>
        <header className={styles.header}>
          <div className={styles.leftsideControls}>
            {!hideSearch && (
              <>
                <TextInput
                  placeholder="Search for a product..."
                  type="search"
                  name="sku-search"
                  id="sku-search"
                  testId="sku-search"
                  value={search}
                  onChange={(event) => this.setSearch((event.target as HTMLInputElement).value)}
                />
                <SearchIcon variant="muted" />
              </>
            )}
            {!!pagination.total && (
              <span className={styles.total}>
                Total results: {pagination.total.toLocaleString()}
              </span>
            )}
          </div>
          <div className={styles.rightsideControls}>
            <ProductSelectionList products={selectedProducts} selectProduct={this.selectProduct} />
            <Button
              className={styles.saveBtn}
              variant="primary"
              onClick={() => this.props.sdk.close(selectedSKUs)}
              isDisabled={selectedSKUs.length === 0}
            >
              {makeSaveBtnText(selectedSKUs, skuType)}
            </Button>
          </div>
        </header>
        <section className={styles.body}>
          <ProductList
            products={products}
            selectProduct={this.selectProduct}
            selectedSKUs={selectedSKUs}
          />
          {!infiniteScrollingPaginationMode && products.length > 0 && (
            <Paginator
              activePage={this.state.activePage}
              className={styles.paginator}
              pageCount={pageCount}
              setActivePage={this.setActivePage}
            />
          )}
          {infiniteScrollingPaginationMode && pagination.hasNextPage && (
            <Button
              className={styles.loadMoreButton}
              variant="transparent"
              testId="infinite-scrolling-pagination"
              onClick={this.loadMoreProducts}
              isFullWidth
            >
              Load more
            </Button>
          )}
        </section>
      </>
    );
  }
}
