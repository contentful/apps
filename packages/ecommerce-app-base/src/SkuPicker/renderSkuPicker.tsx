import React from 'react';
import { DialogExtensionSDK } from '@contentful/app-sdk';
import { render } from 'react-dom';
import { SkuPicker } from './SkuPicker';
import { MakeSaveBtnTextFn, ProductPreviewsFn, ProductsFn } from '../interfaces';

interface Props {
  sdk: DialogExtensionSDK;
  fetchProductPreviews: ProductPreviewsFn;
  fetchProducts: ProductsFn;
  searchDelay?: number;
  skuType?: string;
  makeSaveBtnText?: MakeSaveBtnTextFn;
  hideSearch?: boolean;
}

export function renderSkuPicker(
  elementId: string,
  {
    sdk,
    fetchProductPreviews,
    fetchProducts,
    searchDelay,
    skuType,
    makeSaveBtnText,
    hideSearch,
  }: Props
): void {
  const root = document.getElementById(elementId);

  render(
    <SkuPicker
      sdk={sdk}
      fetchProductPreviews={fetchProductPreviews}
      fetchProducts={fetchProducts}
      searchDelay={searchDelay}
      skuType={skuType}
      makeSaveBtnText={makeSaveBtnText}
      hideSearch={hideSearch}
    />,
    root
  );
}
