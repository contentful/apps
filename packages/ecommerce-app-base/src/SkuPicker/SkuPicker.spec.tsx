import * as React from 'react';
import { merge } from 'lodash';
import { fireEvent, configure, render, cleanup } from '@testing-library/react';
import { Props, SkuPicker } from './SkuPicker';
import { productsList } from '../__mocks__';
import { DialogAppSDK } from '@contentful/app-sdk';
import { Integration, ProductsFn } from '../types';

configure({
  testIdAttribute: 'data-test-id',
});

const renderComponent = async (props: Props) => {
  const component = render(<SkuPicker {...props} />);
  // wait for data to load and render
  await component.findByText('Dress Twin-Set rose');
  component.getAllByTestId('image').forEach((img) => fireEvent(img, new Event('load')));
  return component;
};

jest.mock('react-sortable-hoc', () => ({
  SortableContainer: (x: unknown) => x,
  SortableElement: (x: unknown) => x,
  SortableHandle: (x: unknown) => x,
}));

describe('SkuPicker', () => {
  let defaultProps: Props;
  beforeEach(() => {
    defaultProps = {
      sdk: {
        parameters: {
          installation: {},
          invocation: {
            fieldType: 'Symbol',
            fieldValue: [],
          },
        },
        close: jest.fn(),
        notifier: {
          success: jest.fn(),
          error: jest.fn(),
        },
      } as unknown as DialogAppSDK,
      fetchProductPreviews: jest.fn((skus) =>
        productsList.filter((preview) => skus.includes(preview.sku))
      ) as unknown as Integration['fetchProductPreviews'],
      fetchProducts: jest.fn(() => ({
        pagination: {
          count: 3,
          limit: 20,
          offset: 0,
          total: 3,
        },
        products: productsList,
      })) as unknown as ProductsFn,
    };
  });
  afterEach(cleanup);

  it('should render basic search successfully with no products selected', async () => {
    const { getByTestId, queryByTestId } = await renderComponent(defaultProps);
    expect(getByTestId('sku-search')).toBeInTheDocument();
    expect(queryByTestId('search-by-sku')).not.toBeInTheDocument();
  });

  it('should not render search when hideSearch is true', async () => {
    const { queryByTestId } = await renderComponent({ ...defaultProps, hideSearch: true });
    expect(queryByTestId('sku-search')).not.toBeInTheDocument();
    expect(queryByTestId('search-by-sku')).not.toBeInTheDocument();
  });

  it('should render search by sku option if showSearchBySkuOption is true', async () => {
    const { getByTestId } = await renderComponent({ ...defaultProps, showSearchBySkuOption: true });
    expect(getByTestId('search-by-sku')).toBeInTheDocument();
  });

  it('should render custom placeholder text in search box when makeSearchPlaceholderText exists', async () => {
    const makeSearchPlaceholderText = jest.fn(() => 'My custom placeholder text');
    const { getByPlaceholderText } = await renderComponent({
      ...defaultProps,
      makeSearchPlaceholderText,
    });
    expect(getByPlaceholderText('My custom placeholder text')).toBeInTheDocument();
  });

  it('should render custom help text under search box when makeSearchHelpText exists', async () => {
    const makeSearchHelpText = jest.fn(() => 'My custom help text');
    const { getByText } = await renderComponent({
      ...defaultProps,
      makeSearchHelpText,
    });
    expect(getByText('My custom help text')).toBeInTheDocument();
  });

  describe('when it has infinite scrolling mode pagination', () => {
    it('should render the "Load more" text link if there is a next page', async () => {
      const { findByTestId } = await renderComponent({
        ...defaultProps,
        fetchProducts: jest.fn(() => ({
          pagination: {
            hasNextPage: true,
          },
          products: productsList.slice(0, 2),
        })) as unknown as ProductsFn,
      });
      expect(await findByTestId('infinite-scrolling-pagination')).toBeTruthy();
    });

    it('should not render the "Load more" text link if there is no next page', async () => {
      const { queryByTestId } = await renderComponent({
        ...defaultProps,
        fetchProducts: jest.fn(() => ({
          pagination: {
            hasNextPage: false,
          },
          products: productsList.slice(0, 2),
        })) as unknown as ProductsFn,
      });
      expect(queryByTestId('infinite-scrolling-pagination')).toBeNull();
    });
  });

  describe('when it is operating on a field of type Symbol', () => {
    it('should allow the user to select only one product', async () => {
      const { queryByTestId, findByTestId } = await renderComponent(defaultProps);
      const productA = await findByTestId(`product-preview-${productsList[1].sku}`);
      const productB = await findByTestId(`product-preview-${productsList[2].sku}`);
      productA.click();
      productB.click();
      expect(queryByTestId(`selection-preview-${productsList[1].sku}`)).toBeNull();
      expect(await findByTestId(`selection-preview-${productsList[2].sku}`)).toBeTruthy();
    });
  });

  describe('when it is operating on a field of type Symbol[]', () => {
    it('should allow the user to select multiple products', async () => {
      const { findByTestId } = await renderComponent({
        ...defaultProps,
        sdk: merge({}, defaultProps.sdk, {
          parameters: { invocation: { fieldType: 'Array' } },
        }),
      });
      const productA = await findByTestId(`product-preview-${productsList[1].sku}`);
      const productB = await findByTestId(`product-preview-${productsList[2].sku}`);
      productA.click();
      productB.click();
      expect(await findByTestId(`selection-preview-${productsList[1].sku}`)).toBeTruthy();
      expect(await findByTestId(`selection-preview-${productsList[2].sku}`)).toBeTruthy();
    });
  });
});
