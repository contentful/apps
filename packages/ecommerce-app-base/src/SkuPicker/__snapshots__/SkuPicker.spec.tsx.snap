// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SkuPicker should render successfully with no products selected 1`] = `
<div>
  <header
    class="css-zv97dc"
  >
    <div
      class="css-kpva4h"
    >
      <div
        class="TextInput__TextInput___36-K- TextInput__TextInput--medium___1bR2D"
      >
        <input
          class="TextInput__TextInput__input___27vDB a11y__focus-border--default___60AXp"
          data-test-id="sku-search"
          id="sku-search"
          name="sku-search"
          placeholder="Search for a product..."
          type="search"
          value=""
        />
      </div>
      <svg
        class="Icon__Icon___38Epv Icon__Icon--small___1yGZK Icon__Icon--muted___3egnD"
        data-test-id="cf-ui-icon"
        height="1em"
        viewBox="0 0 24 24"
        width="1em"
      >
        <path
          d="M15.5 14h-.79l-.28-.27A6.471 6.471 0 0016 9.5 6.5 6.5 0 109.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
        />
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
      </svg>
      <span
        class="css-1x7yppz"
      >
        Total results: 
        3
      </span>
    </div>
    <div
      class="css-kb3y9n"
    >
      <div
        class="css-k008qs"
      />
      <button
        class="Button__Button___1ZfFj css-1kr4fmq Button__Button--primary___JImeO Button__Button--disabled___1E20M"
        data-test-id="cf-ui-button"
        disabled=""
        type="button"
      >
        <span
          class="TabFocusTrap__TabFocusTrap___39Vty Button__Button__inner-wrapper___3qrNC"
          tabindex="-1"
        >
          <span
            class="Button__Button__label___3tcOj"
          >
            Save products
          </span>
        </span>
      </button>
    </div>
  </header>
  <section
    class="css-1gzkqoh"
  >
    <div
      class="css-almiqb"
    >
      <div
        class="css-jfu8ng"
      >
        <div
          aria-checked="false"
          class="css-16t4lg0 "
          data-test-id="product-preview-M0E20130820E90Z"
          role="switch"
          tabindex="-1"
        >
          <div
            class="css-1dtkwc5"
          >
            <img
              alt="product preview"
              class="css-1ddetr"
              data-test-id="image"
              src="https://example.com/commercetools-example/products/picture-1.jpg"
              style="display: block;"
            />
            <img
              alt="check"
              class="css-1imdisq"
              src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iNDQiIGhlaWdodD0iNDQiIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyI+PGcgZmlsbD0iIzMwNzJCRSIgZmlsbC1ydWxlPSJldmVub2RkIj48Y2lyY2xlIGZpbGwtb3BhY2l0eT0iLjEiIGN4PSIyMiIgY3k9IjIyIiByPSIyMiIvPjxwYXRoIGQ9Ik0zMS42ODkgMTYuNjE0TDI5LjU1NSAxNC41IDE5LjAyNCAyNS4yNzVsLTQuNjItNC40MDYtMi4wOTIgMi4xNTMgNi43NSA2LjQ3OCAxMi42MjctMTIuODg2eiIgZmlsbC1ydWxlPSJub256ZXJvIi8+PC9nPjwvc3ZnPg=="
            />
          </div>
          <p
            class="css-182natc"
          >
            Dress Twin-Set rose
          </p>
          <p
            class="css-1bgao96"
          >
            M0E20130820E90Z
          </p>
        </div>
      </div>
      <div
        class="css-jfu8ng"
      >
        <div
          aria-checked="false"
          class="css-16t4lg0 "
          data-test-id="product-preview-A0E2300FX102203"
          role="switch"
          tabindex="-1"
        >
          <div
            class="css-1dtkwc5"
          >
            <img
              alt="product preview"
              class="css-1ddetr"
              data-test-id="image"
              src="https://example.com/commercetools-example/products/picture-2.jpg"
              style="display: block;"
            />
            <img
              alt="check"
              class="css-1imdisq"
              src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iNDQiIGhlaWdodD0iNDQiIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyI+PGcgZmlsbD0iIzMwNzJCRSIgZmlsbC1ydWxlPSJldmVub2RkIj48Y2lyY2xlIGZpbGwtb3BhY2l0eT0iLjEiIGN4PSIyMiIgY3k9IjIyIiByPSIyMiIvPjxwYXRoIGQ9Ik0zMS42ODkgMTYuNjE0TDI5LjU1NSAxNC41IDE5LjAyNCAyNS4yNzVsLTQuNjItNC40MDYtMi4wOTIgMi4xNTMgNi43NSA2LjQ3OCAxMi42MjctMTIuODg2eiIgZmlsbC1ydWxlPSJub256ZXJvIi8+PC9nPjwvc3ZnPg=="
            />
          </div>
          <p
            class="css-182natc"
          >
            Michael Kors – Shopper “Jet Set Travel”
          </p>
          <p
            class="css-1bgao96"
          >
            A0E2300FX102203
          </p>
        </div>
      </div>
      <div
        class="css-jfu8ng"
      >
        <div
          aria-checked="false"
          class="css-16t4lg0 "
          data-test-id="product-preview-M0E21300900DZN7"
          role="switch"
          tabindex="-1"
        >
          <div
            class="css-1dtkwc5"
          >
            <img
              alt="product preview"
              class="css-1ddetr"
              data-test-id="image"
              src="https://example.com/commercetools-example/products/picture-3.jpg"
              style="display: block;"
            />
            <img
              alt="check"
              class="css-1imdisq"
              src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iNDQiIGhlaWdodD0iNDQiIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyI+PGcgZmlsbD0iIzMwNzJCRSIgZmlsbC1ydWxlPSJldmVub2RkIj48Y2lyY2xlIGZpbGwtb3BhY2l0eT0iLjEiIGN4PSIyMiIgY3k9IjIyIiByPSIyMiIvPjxwYXRoIGQ9Ik0zMS42ODkgMTYuNjE0TDI5LjU1NSAxNC41IDE5LjAyNCAyNS4yNzVsLTQuNjItNC40MDYtMi4wOTIgMi4xNTMgNi43NSA2LjQ3OCAxMi42MjctMTIuODg2eiIgZmlsbC1ydWxlPSJub256ZXJvIi8+PC9nPjwvc3ZnPg=="
            />
          </div>
          <p
            class="css-182natc"
          >
            Shirt “Jenny“ Polo Ralph Lauren white
          </p>
          <p
            class="css-1bgao96"
          >
            M0E21300900DZN7
          </p>
        </div>
      </div>
    </div>
    <div
      class="css-1twx1i5"
    >
      <button
        class="Button__Button___1ZfFj css-fedt8m Button__Button--muted___2Wair Button__Button--disabled___1E20M"
        data-test-id="cf-ui-button"
        disabled=""
        type="button"
      >
        <span
          class="TabFocusTrap__TabFocusTrap___39Vty Button__Button__inner-wrapper___3qrNC"
          tabindex="-1"
        >
          <span
            class="Button__Button__label___3tcOj"
          >
            <img
              alt="right"
              class="css-1t0k538"
              src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMTgiIGhlaWdodD0iMTgiIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyI+PGcgZmlsbD0ibm9uZSIgZmlsbC1ydWxlPSJldmVub2RkIj48cGF0aCBkPSJNMCAwaDE4djE4SDB6Ii8+PHBhdGggZD0iTTEzLjIyMSA0LjVMOC43MjEgOWw0LjUgNC41IDEuMDU4LTEuMDU3TDEwLjg0NCA5bDMuNDM1LTMuNDQyTDEzLjIyIDQuNXptLTUgMEwzLjcyMSA5bDQuNSA0LjUgMS4wNTgtMS4wNTdMNS44NDQgOWwzLjQzNS0zLjQ0Mkw4LjIyIDQuNXoiIGZpbGw9IiMwMDAiIGZpbGwtcnVsZT0ibm9uemVybyIvPjwvZz48L3N2Zz4="
            />
          </span>
        </span>
      </button>
      <button
        class="Button__Button___1ZfFj css-fedt8m Button__Button--muted___2Wair Button__Button--disabled___1E20M"
        data-test-id="cf-ui-button"
        disabled=""
        type="button"
      >
        <span
          class="TabFocusTrap__TabFocusTrap___39Vty Button__Button__inner-wrapper___3qrNC"
          tabindex="-1"
        >
          <svg
            class="Icon__Icon___38Epv Icon__Icon--small___1yGZK Icon__Icon--secondary___1ztcw Button__Button__icon___2YX5-"
            data-test-id="cf-ui-icon"
            height="1em"
            viewBox="0 0 24 24"
            width="1em"
          >
            <path
              d="M15.3738 5.29289C15.7643 5.68342 15.7643 6.31658 15.3738 6.70711L10.0809 12L15.3738 17.2929C15.7643 17.6834 15.7643 18.3166 15.3738 18.7071C14.9833 19.0976 14.3501 19.0976 13.9596 18.7071L7.95956 12.7071C7.56904 12.3166 7.56904 11.6834 7.95956 11.2929L13.9596 5.29289C14.3501 4.90237 14.9833 4.90237 15.3738 5.29289Z"
            />
          </svg>
        </span>
      </button>
      <button
        class="Button__Button___1ZfFj css-fedt8m Button__Button--primary___JImeO"
        data-test-id="active"
        type="button"
      >
        <span
          class="TabFocusTrap__TabFocusTrap___39Vty Button__Button__inner-wrapper___3qrNC"
          tabindex="-1"
        >
          <span
            class="Button__Button__label___3tcOj"
          >
            1
          </span>
        </span>
      </button>
      <button
        class="Button__Button___1ZfFj css-fedt8m Button__Button--muted___2Wair Button__Button--disabled___1E20M"
        data-test-id="cf-ui-button"
        disabled=""
        type="button"
      >
        <span
          class="TabFocusTrap__TabFocusTrap___39Vty Button__Button__inner-wrapper___3qrNC"
          tabindex="-1"
        >
          <svg
            class="Icon__Icon___38Epv Icon__Icon--small___1yGZK Icon__Icon--secondary___1ztcw Button__Button__icon___2YX5-"
            data-test-id="cf-ui-icon"
            height="1em"
            viewBox="0 0 24 24"
            width="1em"
          >
            <path
              d="M8.62623 18.7071C8.2357 18.3166 8.2357 17.6834 8.62623 17.2929L13.9191 12L8.62623 6.70711C8.2357 6.31658 8.2357 5.68342 8.62623 5.29289C9.01675 4.90237 9.64992 4.90237 10.0404 5.29289L16.0404 11.2929C16.431 11.6834 16.431 12.3166 16.0404 12.7071L10.0404 18.7071C9.64992 19.0976 9.01675 19.0976 8.62623 18.7071Z"
            />
          </svg>
        </span>
      </button>
      <button
        class="Button__Button___1ZfFj css-fedt8m Button__Button--muted___2Wair Button__Button--disabled___1E20M"
        data-test-id="cf-ui-button"
        disabled=""
        type="button"
      >
        <span
          class="TabFocusTrap__TabFocusTrap___39Vty Button__Button__inner-wrapper___3qrNC"
          tabindex="-1"
        >
          <span
            class="Button__Button__label___3tcOj"
          >
            <img
              alt="right"
              class="css-11hf54p"
              src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMTgiIGhlaWdodD0iMTgiIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyI+PGcgZmlsbD0ibm9uZSIgZmlsbC1ydWxlPSJldmVub2RkIj48cGF0aCBkPSJNMCAwaDE4djE4SDB6Ii8+PHBhdGggZD0iTTEzLjIyMSA0LjVMOC43MjEgOWw0LjUgNC41IDEuMDU4LTEuMDU3TDEwLjg0NCA5bDMuNDM1LTMuNDQyTDEzLjIyIDQuNXptLTUgMEwzLjcyMSA5bDQuNSA0LjUgMS4wNTgtMS4wNTdMNS44NDQgOWwzLjQzNS0zLjQ0Mkw4LjIyIDQuNXoiIGZpbGw9IiMwMDAiIGZpbGwtcnVsZT0ibm9uemVybyIvPjwvZz48L3N2Zz4="
            />
          </span>
        </span>
      </button>
    </div>
  </section>
</div>
`;
